### ISSUE TITLE
Can user-defined annotations on a field be accessed during mapping ?

### ISSUE THREAD

**Issue by amitga-rnd** [mapstruct/mapstruct#1470]
Hello.
I started playing around with MapStruct, and think its an awesome framework!
But, I have a certain use case, that I haven't found a solution for yet:

Lets say I have some user-defined annotations on a field of a class. Can I access this annotation somehow during the mapping/conversion to another class, in order to affect the conversion of the value of the field ?

For example:
```
public class PersonEntity extends Entity {
    public String birthdate;

    @GeoConvert(format="WKT")
    public String addressGeo;

    public Name name;

    public String gender;
}
```
When I convert `PersonEntity` to `Person`, I want to read the annotations on `addressGeo` field (of source class), so that I can pass the field's value, and the annotation's value to a converter (reverse geocoding in this case)  and set the target's `address` field accordingly.

This is just an example, I have some other annotations, that work in a similar manner.

I guess I could do this manually, looking at each entity's fields, and mapping it accordingly, but I preffer not to do this, as I'm not the one writing the entities (I get them from a jar file written by another team), and I don't want to change the mapping, every time I receive an update of the entities jar.
Another option is to user `@AfterMapping`, and then use reflection to find annotations on the source object's class, but this probably wouldn't be very performant (and kinda breaks the fact the MapStruct doesn't user reflection in runtime, which I like :-))

Any other ideas? Could this be implemented somehow in the framework (future feature)?
Thanks in advance.

**Comment by sjaakd** [comment]
Did you have a look at qualifiers? That might solve your use case.

**Comment by amitga-rnd** [comment]
Yes I have, but it doesn't exactly solve my use case.
I need to know the value of a parameter in the annotation (i.e. `format="WKT"` in the  @GeoConvert annotation) in order to properly translate the values of that field.
It would be nice if this could be extracted and then sent to a conversion method that I have.

I have other annotations, where I need the of a parameter in them in order to map, or translate the value correctly.

**Comment by sjaakd** [comment]
How many formats are you expecting? Would an annotation `@GeoConvertWKT` also be an option?

**Comment by sjaakd** [comment]
ah.. esri stuff. are you dealing with object graphs? or single mapping methods?

**Comment by sjaakd** [comment]
I was thinking, we have context objects.. (annotated `@Context`).. perhaps they are somehow usable...

**Comment by sjaakd** [comment]
So.. ideas start to form a bit.. And I must say.. I'm not entirely sure if this somehow achievable... But suppose you would have a method:

**Comment by amitga-rnd** [comment]
Yes, I think that could be a solution!
Is it possible currently with MapStruct or does your examole need development?

**Comment by amitga-rnd** [comment]
OK. I am not familiar with annotation processors, but I guess I could look into it.
Could you please focus me on what parts of the code are relevant, if I want to implement it?

BTW, thanks for your prompt responses and cooperation :+1:

**Comment by sjaakd** [comment]
`BeanMappingMethod` and `PropertyMappingMethod` are good candidates to start.. We use Freemarker as tool to generate code  `BeanMappingMethod` and `PropertyMappingMethod` both have a backing bean. I am going to start with implementing object factories on the `Context` object soon (so probably will be working on the same code.

I was thinking: with the current state of MapStruct you could also use the `@BeforeMapping` in combination to get access to a context object. In the implementation of `@BeforeMapping` you could do more or less the same as we are discussing above.. So you hand the context object to the `@BeforeMapping` and colllect annotations on the properties of the bean you are dealing with and store them in the `@Context` annotated property.

**Comment by amitga-rnd** [comment]
>  was thinking: with the current state of MapStruct you could also use the @BeforeMapping in combination to get access to a context object. In the implementation of @BeforeMapping you could do more or less the same as we are discussing above.. So you hand the context object to the @BeforeMapping and colllect annotations on the properties of the bean you are dealing with and store them in the @Context annotated property.

I will look into it. I was trying to prevent use of reflection as much as possible (reading annotated fields of a class), but I think it might not be that bad. There is a sort of a stigma as to using reflection (even indirectly), but I guess the JIT/JVM is smart enough to optimize these calls. Java isn't real-time anyways, so maybe my worries about using such methods are exaggerated.

**Comment by sjaakd** [comment]
@amitga-rnd .. I've added an example with the current code that might interest you. I need to strip it from another Joda example.. Will do that soon. But have a look [here](https://github.com/mapstruct/mapstruct-examples/pull/43)..

**Comment by sjaakd** [comment]
If we want to solve this in MapStruct, the context is perhaps not so usable. It has the scope of the bean mapping method, not of an individual mapping. This is apparent in the example were I need a map, indexed by the target setter name. Better would be to add some annotations and populate those when available. E.g. `target map( SourceType source, @TargetAnnotation Treatment treatment, @TargetAnnotation Treatment2 treatment2, @SourceAnnotation Something something)` 

@filiphr came with a nice article how to 'mock' an annotation with an anonymous inner class. See [here](https://stackoverflow.com/questions/16299717/how-to-create-an-instance-of-an-annotation/16326389#16326389).

**Comment by amitga-rnd** [comment]
Yes, the latter prototype with the annotations is more like I meant originally. 
I tried to implement the use case using the context (a bit different than your example), but still its not so elegant (bean mapping vs. Field mapping,  as you mentioned)

**Comment by sjaakd** [comment]
The example solution is usable right away. 

I'm not sure whether one of us has time for implementing the 'MapStruct` solution. Its possibly quite some work. Also, I would like to discuss the change in MapStruct with some of of the other contributors. @filiphr, can we take this one in one of our hangouts and discuss it with @gunnarmorling and @agudian?.
